### DISCUSSION TITLE
Mac80211 Patch - There is another way to patch it without build a new Kernel?

### DISCUSSION THREAD

**Discussion by Special-Niewbie** [thefloppydriver/pc2drc#8] [Q&A] [answered]
Hello @thefloppydriver ,

in the meantime the things keep going on, I would like to share and ask you this as per title.

There is another way to patch the Mac80211 (like as a normal software installation of the package), example in the existing Linux Kernel without mess up it if something wrong ? 

I ask you this, because if it's between us as Advanced Users it's fine to build, test and so on; but I'm thinking for more normal people that would like to enjoy this milestone software and ask them to build them own kernel (even if your steps are amazingly automated) sometimes if something happen during the process could mess up with the own Linux installation.

Also, once everything will be done from my side, if I need to install it in my principal Linux installation, if it will get some problem on it - Linux as you can understand to set and well personalized it take a lot of time to set -  so, if something goes wrong and i need to reinstall everything.....

Any suggestion about?

**Comment by thefloppydriver** [comment, marked as answer]
Short answer: Yes, because step-1-kernel-patch doesn't build a new kernel, it downloads kernel sources and builds a kernel module and loads that into the currently running kernel :)

Long answer:
stage-1-kernel-patch doesn't build a new kernel. It used to but now it doesn't. The old script that fully rebuilds a new kernel is `./kernel-patch-files/build-patched-kernel.sh`. Also don't worry about it making your system unusable. I took great care to make sure that neither script would cause any damage if it failed. I'm confident in saying that they're both completely safe to run.
If anything fails, the script will try to do it a different way. If that fails then the script will error out, cleanup, and exit. It's very safe, and I use it on all three of my linux installations.

Below is *exactly* how stage-1-kernel-patch works.


1) Do initialization sanity checks (check if user ran the script with `sudo -E <script name>`; check if script is in pc2drc folder; check if there are any wireless interfaces attached to the system; check for active internet connection)

2) Install required packages via apt-get

3) Get currently running kernel version.

4) If kernel sources already downloaded, skip download; else, pull virgin kernel sources matching currently running kernel version from https://cdn.kernel.org/ (if kernel verison == 5.12.11, the url would be https://cdn.kernel.org/pub/linux/kernel/v5.x/linux-5.12.11.tar.xz)

5) If kernel sources already extracted, skip extraction; else, extract kernel sources (I will be referring to this folder as the kernel source tree from now on.)

6) Copy the current kernel config (/boot/config-[kernel version]) to the top-level directory of the kernel source tree

7) Patch the mac80211 folder in kernel source tree (if kernel version > 5.11, use new patch; if kernel version <= 5.11, use old patch)

8) Configure kernel source tree

9) Copy the patched mac80211 folder from the kernel source tree to `/usr/src/drc-mac80211-0.1.0`

10) If drc-mac80211-0.1.0 already installed, remove it

11) Make and install drc-mac80211 kernel module.

12) Test if drc-mac80211 is installed

13) If drc-mac80211 did NOT install successfully:
    remove downloaded and extracted kernel sources
    notify user that drc-mac80211 could not be installed using the above methods
    create folder titled "PUT_KERNEL_SOURCES_IN_THIS_FOLDER" in the pc2drc directory
    If the current running linux distro is Ubuntu:
        Check if we've edited the apt-sources file (we left a signature on each line we edited, check for that)
        If we haven't edited apt-sources, put links to Ubuntu sources in apt-sources
        Download Ubuntu kernel sources to `./PUT_KERNEL_SOURCES_IN_THIS_FOLDER`
    Else if the current running linux distro is not Ubuntu:
        Instruct users to get their distro's kernel sources and put them in `./PUT_KERNEL_SOURCES_IN_THIS_FOLDER`
    Try to find kernel sources within `./PUT_KERNEL_SOURCES_IN_THIS_FOLDER`
    If not found kernel sources, instruct user to put kernel sources in `./PUT_KERNEL_SOURCES_IN_THIS_FOLDER` until they do it right.
    If found kernel sources, do steps 6-12 using the kernel sources from `./PUT_KERNEL_SOURCES_IN_THIS_FOLDER`, then delete the folder.
    If kernel patch not installed successfully, generate bug report and exit.
    Else if kernel patch installed successfully, continue.

14) Unload and load all running kernel modules that depend on mac80211 so that the modules load the new, patched module.

15) If there were no modules using mac80211, then it is extremely likely that the user does not have any wireless interfaces that use mac80211. Therefore, inform the user that they do not have any compatible wireless interfaces.

16) If `/sys/class/net/$(ip link show | grep -o -m1 "\w*wl\w*")/tsf` exists, then the user has a wireless interface that uses mac80211 AND the kernel patch is installed and loaded. Success. Delete all linux source files and directories in `./kernel-patch-files` and exit.

17) If `/sys/class/net/$(ip link show | grep -o -m1 "\w*wl\w*")/tsf` does NOT exist:
    If the user has no mac80211 compatible wireless interfaces:
        inform user that the kernel patch installed, but it isn't working. Also inform user that the reason the kernel patch "isn't working" is probably because they don't have any mac80211 compatible interfaces and link them to the pc2drc readme file.
    Else If the user DOES have a mac80211 compatible wireless interface:
        The script failed entirely. Instruct user to reboot and run the script again and if that doesn't work to run `./kernel-patch-files/build-patched-kernel.sh` (last resort)

**Reply by Special-Niewbie** [reply]
Wow, thanks a lot for such a thorough answer, I really appreciate it!!! 😊

**Reply by thefloppydriver** [reply]
@Special-Niewbie No problem! I'd want the same done to me if someone was messing around with system files on my machine :)
